### PR TITLE
fix-sklearn-autolog-without-matplotlib

### LINKED ISSUE #5994: [BUG] sklearn.autolog fails without matplotlib installed
### Willingness to contribute

Yes. I can contribute a fix for this bug independently.

### MLflow version

1.26.1.dev0

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: MacOS (with M1 chip)
- **Python version**: 3.8.13
- **yarn version, if running the dev UI**:

### Describe the problem

Since release 1.26.0, mlflow.sklearn.autolog fails to log experiments if matplotlib is not installed.

The issue is that you import matplotlib outside of the try block on line 489 in mlflow/mlflow/sklearn/utils.py.

Proposed fix: move matplotlib code into the try/except block (only place where it is used anyways).

### Tracking information

_No response_

### Code to reproduce issue

def tree_trainer_with_mlflow(
        X_train: np.ndarray,
        y_train: np.ndarray,
    ):
    mlflow.sklearn.autolog()  # log all model hparams and metrics to MLflow
    model = DecisionTreeClassifier()
    model.fit(X_train, y_train)
    return model

### Other info / logs

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by fa9r**
Signed-off-by: Felix Altenberger <felix@zenml.io>

## Related Issues/PRs

Close #5994

## What changes are proposed in this pull request?

Moved matplotlib import in sklearn/utils.py into try block to prevent mlflow.sklearn.autolog() from breaking when matplotlib is not installed.

## How is this patch tested?

Manually (by testing with our own mlflow integration examples at [zenml](https://github.com/zenml-io/zenml))

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/sklearn/utils.py [hunk 1]
```diff
@@ -524,19 +523,24 @@ def _log_specialized_estimator_content(
...
+                        "figure.autolayout": True,
+                        "font.size": 8,
+                    }
+                    with matplotlib.rc_context(_matplotlib_config):
+                        display = artifact.function(**artifact.arguments)
+                        display.ax_.set_title(artifact.title)
+                        artifact_path = "{}.png".format(artifact.name)
+                        filepath = tmp_dir.path(artifact_path)
+                        display.figure_.savefig(fname=filepath, format="png")
+                        plt.close(display.figure_)
```
Comment (harupy) [inline mlflow/sklearn/utils.py:543]:
```suggestion
                        plt.close(display.figure_)
                    except ImportError:
                        # warning message
                        break
```

@fa9r Thanks for the PR! Can we break if we fail to import `matplotlib`?

Comment (fa9r) [reply mlflow/sklearn/utils.py:543]: Added break for ImportErrors in Exception block below

Comment (harupy) [reply mlflow/sklearn/utils.py:543]: Can we separate except branches?

```python
try:
   ...
except ImportError as ie:
   ...
except Exception as e:
    ...
```

Comment (fa9r) [reply mlflow/sklearn/utils.py:543]: Sure. That is indeed more readable, good catch.

**Comment by harupy** [comment]
I'm looking at the implementation of `_log_specialized_estimator_content`. Can we attempt to import `matplotlib` before the loop like this?

```diff
diff --git a/mlflow/sklearn/utils.py b/mlflow/sklearn/utils.py
index f8720a793..3f46717b2 100644
--- a/mlflow/sklearn/utils.py
+++ b/mlflow/sklearn/utils.py
@@ -523,17 +523,21 @@ def _log_specialized_estimator_content(
             _logger.warning(msg)
             return
 
+        try:
+            import matplotlib
+            import matplotlib.pyplot as plt
+        except ImportError:
+            # log warning message here
+            return metrics
+
         with TempDir() as tmp_dir:
+            _matplotlib_config = {
+                "savefig.dpi": 175,
+                "figure.autolayout": True,
+                "font.size": 8,
+            }
             for artifact in artifacts:
                 try:
-                    import matplotlib
-                    import matplotlib.pyplot as plt
-
-                    _matplotlib_config = {
-                        "savefig.dpi": 175,
-                        "figure.autolayout": True,
-                        "font.size": 8,
-                    }
                     with matplotlib.rc_context(_matplotlib_config):
                         display = artifact.function(**artifact.arguments)
                         display.ax_.set_title(artifact.title)
@@ -541,9 +545,6 @@ def _log_specialized_estimator_content(
                         filepath = tmp_dir.path(artifact_path)
                         display.figure_.savefig(fname=filepath, format="png")
                         plt.close(display.figure_)
-                except ImportError as e:
-                    _log_warning_for_artifacts(artifact.name, artifact.function, e)
-                    break
                 except Exception as e:
                     _log_warning_for_artifacts(artifact.name, artifact.function, e)
```

File: mlflow/ml-package-versions.yml [hunk 2]
```diff
@@ -21,6 +21,10 @@ sklearn:
     run: |
       pytest tests/sklearn/test_sklearn_autolog.py
 
+      # Ensure sklearn autologging works without matplotlib
+      pip uninstall -q -y matplotlib
+      pytest tests/sklearn/test_sklearn_autolog_without_matplotlib.py
```
Comment (harupy) [inline mlflow/ml-package-versions.yml:26]: @fa9r Added a test here.

Comment (harupy) [reply mlflow/ml-package-versions.yml:26]: It worked: https://github.com/mlflow/mlflow/runs/6689715702?check_suite_focus=true#step:12:253

**Review by harupy — APPROVED** [review]
LGTM :)

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**